manager/common/license.go

```diff
@@ -0,0 +1,192 @@
+package license
+
+import (
+	astoreServer "github.com/enfabrica/enkit/astore/server/astore"
+	rpc_license "github.com/enfabrica/enkit/manager/rpc"
+	"google.golang.org/grpc/metadata"
+	"io"
+	"log"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+)
+
+type LicenseCounter struct {
+	counterMutex sync.Mutex
+	clientMutex  sync.Mutex
+	clients      map[string]*Client  // client connection metadata
+	licenses     map[string]*License // current and total licenses available per vendor
+	queue        map[string][]string // queue of jobs waiting to be handled
+}
+
+func (c *LicenseCounter) increment(vendor string, num int, hash string) int {
+	c.counterMutex.Lock()
+	defer c.counterMutex.Unlock()
+	if c.licenses[vendor].Used < c.licenses[vendor].Total {
+		log.Printf("License acquired by %s \n", hash)
+		c.licenses[vendor].Used += num
+	} else {
+		log.Fatalf("Cannot acquire more licenses than available \n")
```

Comment:
in increment, decrement, etc:
1) invert the if, remove the else.
2) Don't go Fatal, just log.Errorf

Example:

```
if c.licenses[vendor].Used >= c.licenses[vendor].Total {
  log.Errorf(... cannot acquire ...)
  return;
}

c.licenses[vendor].Used += num
```

if you crash and burn with the Fatal, all builds in progress will fail/start failing. Further, docker will restart the server, and make debugging harder, making everything flaky.

Might as well "log the error" and keep going. Will be easier to debug, and will make the system more stable.